Ok, there is a test failure on Windows:

```
test_pool_worker_lifetime (test.test_multiprocessing_spawn.WithProcessesTestPoolWorkerLifetime) ... Exception in thread Thread-908:
Traceback (most recent call last):
  File "C:\projects\cpython\lib\threading.py", line 917, in _bootstrap_inner
    self.run()
  File "C:\projects\cpython\lib\threading.py", line 865, in run
    self._target(*self._args, **self._kwargs)
  File "C:\projects\cpython\lib\multiprocessing\pool.py", line 511, in _handle_workers
    cls._wait_for_updates(sentinels, change_notifier)
  File "C:\projects\cpython\lib\multiprocessing\pool.py", line 494, in _wait_for_updates
    wait(sentinels, timeout=timeout)
  File "C:\projects\cpython\lib\multiprocessing\connection.py", line 878, in wait
    ready_handles = _exhaustive_wait(waithandle_to_obj.keys(), timeout)
  File "C:\projects\cpython\lib\multiprocessing\connection.py", line 810, in _exhaustive_wait
    res = _winapi.WaitForMultipleObjects(L, False, timeout)
OSError: [WinError 6] The handle is invalid
```

--- a/Lib/multiprocessing/pool.py
+++ b/Lib/multiprocessing/pool.py
@@ -265,6 +265,8 @@ class Pool(object):
         if self._state == RUN:
             _warn(f"unclosed running multiprocessing pool {self!r}",
                   ResourceWarning, source=self)
+            if getattr(self, '_change_notifier') is not None:
+                self._change_notifier.put(None)
 
     def __repr__(self):
         cls = self.__class__

--- a/Lib/multiprocessing/pool.py
+++ b/Lib/multiprocessing/pool.py
@@ -218,14 +218,8 @@ class Pool(object):
             self_notifier_sentinels = [self._change_notifier._reader]
         else:
             self_notifier_sentinels = []
-        try:
-            worker_sentinels = [worker.sentinel for worker in self._pool]
-        except AttributeError:
-            worker_sentinels = []
 
-        sentinels = [*task_queue_sentinels,
-                     *worker_sentinels,
-                     *self_notifier_sentinels]
+        sentinels = [*task_queue_sentinels, *self_notifier_sentinels]
 
         self._worker_handler = threading.Thread(
             target=Pool._handle_workers,
@@ -490,7 +484,7 @@ class Pool(object):
         return result
 
     @staticmethod
-    def _wait_for_updates(sentinels, change_notifier, timeout=0.1):
+    def _wait_for_updates(sentinels, change_notifier, timeout=None):
         wait(sentinels, timeout=timeout)
         while not change_notifier.empty():
             change_notifier.get()
@@ -506,9 +500,14 @@ class Pool(object):
         # is terminated.
         while thread._state == RUN or (cache and thread._state != TERMINATE):
             cls._maintain_pool(ctx, Process, processes, pool, inqueue,
-                                outqueue, initializer, initargs,
-                                maxtasksperchild, wrap_exception)
-            cls._wait_for_updates(sentinels, change_notifier)
+                               outqueue, initializer, initargs,
+                               maxtasksperchild, wrap_exception)
+
+            worker_sentinels = [worker.sentinel for worker in
+                                pool if hasattr(worker, "sentinel")]
+            current_sentinels = [*worker_sentinels, *sentinels]
+
+            cls._wait_for_updates(current_sentinels, change_notifier)
         # send sentinel to stop workers
         taskqueue.put(None)
         util.debug('worker handler exiting')
@@ -930,5 +929,5 @@ class ThreadPool(Pool):
         for i in range(size):
             inqueue.put(None)
 
-    def _wait_for_updates(self, sentinels, change_notifier, timeout=0.2):
+    def _wait_for_updates(self, sentinels, change_notifier, timeout):
         time.sleep(timeout)

